vendor/github.com/openshift/library-go/pkg/operator/encryption/controllers/migrators/inprocess.go

```diff
@@ -0,0 +1,187 @@
+package migrators
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/pager"
+	"k8s.io/klog"
+)
+
+func NewInProcessMigrator(dynamicClient dynamic.Interface, discoveryClient discovery.ServerResourcesInterface) *InProcessMigrator {
+	return &InProcessMigrator{
+		dynamicClient:   dynamicClient,
+		discoveryClient: discoveryClient,
+		running:         map[schema.GroupResource]*inProcessMigration{},
+	}
+}
+
+// InProcessMigrator runs migration in-process using paging.
+type InProcessMigrator struct {
+	dynamicClient   dynamic.Interface
+	discoveryClient discovery.ServerResourcesInterface
+
+	lock    sync.Mutex
+	running map[schema.GroupResource]*inProcessMigration
+
+	handler cache.ResourceEventHandler
+}
+
+type inProcessMigration struct {
+	stopCh   chan<- struct{}
+	doneCh   <-chan struct{}
+	writeKey string
+	result   *error
+}
+
+func (m *InProcessMigrator) EnsureMigration(gr schema.GroupResource, writeKey string) (finished bool, result error, err error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// finished?
+	migration := m.running[gr]
+	if migration != nil && migration.writeKey == writeKey {
+		if migration.result == nil {
```

Comment:
if all went good (no errors), will `result` be `nil`?

Reply:
added docs